serialize.c

```diff
@@ -0,0 +1,111 @@
+#include "serialize.h"
+
+char binary_to_base64(uint8_t binary) {
+    if(binary <= 25) return binary + 'A';
+    if(binary <= 51) return (binary - 26) + 'a';
+    if(binary <= 61) return (binary - 52) + '0';
+    if(binary == 62) return '&';
+    return '/';
+}
+
+uint8_t base64_to_binary(char base64) {
+    if('A' <= base64 && base64 <= 'Z') return base64 - 'A';
+    if('a' <= base64 && base64 <= 'z') return base64 - 'a' + 26;
+    if('0' <= base64 && base64 <= '9') return base64 - '0' + 52;
+    if(base64 == '&') return 62;
+    return 63;
+}
+
+bool serialize_state(const system_state *state, char *str) {
+    if(!state) return false;
+    if(!str) return false;
```

Comment:
this is a personal best practice thing, but you should test `state != NULL`, not `!state`. C doesn't guarantee that null pointers evaluate to 0.

Reply:
Fixed in ae3c242.